File: 2022/6/solution.py
with open("input.txt") as file:
    text = file.readline()

    start_of_transmission = 0
    
    for i in range(len(text) - 4):
        segment = text[i:i+4]
        buffer = []
        unique = True
        for x in range(4):
            char = text[i + x]
            if char in buffer:
                unique = False
            buffer.append(char)
        if unique:
            print(segment,i + 4)
            start_of_transmission = i + 4
            break
    for i in range(start_of_transmission,len(text) - 14):
        segment = text[i:i+14]
        buffer = []
        unique = True
        for x in range(14):
            char = text[i+x]
            if char in buffer:
                unique = False
            buffer.append(char)
        if unique:
            print(segment,i+14)
            break

File: 2022/6/solution.rs
use std::collections::VecDeque;
use std::fs::read_to_string;

fn main() {
    part_1();
    part_2();
}

fn part_1() {
    let content = read_to_string("input.txt").expect("Could not read input file");
    let mut chariter = content.chars();
    let mut buffer: VecDeque<char> = VecDeque::from([' '; 4]);
    let mut index = 1;

    'chars: loop {
        match (&mut chariter).next() {
            None => {
                index = -1;
                break;
            }
            Some(c) => {
                buffer.pop_front();
                buffer.push_back(c);

                if index < 4 {
                    index += 1;
                    continue;
                }

                for i in 0..4 {
                    for j in (i + 1)..4 {
                        if buffer.get(i).unwrap() == buffer.get(j).unwrap() {
                            index += 1;
                            continue 'chars;
                        }
                    }
                }
                break;
            }
        }
    }

    println!("Part 1: {}", index);
}

fn part_2() {
    let content = read_to_string("input.txt").expect("Could not read input file");
    let mut chariter = content.chars();
    let mut buffer: VecDeque<char> = VecDeque::from([' '; 14]);
    let mut index = 1;

    'chars: loop {
        match (&mut chariter).next() {
            None => {
                index = -1;
                break;
            }
            Some(c) => {
                buffer.pop_front();
                buffer.push_back(c);

                if index < 14 {
                    index += 1;
                    continue;
                }

                for i in 0..14 {
                    for j in (i + 1)..14 {
                        if buffer.get(i).unwrap() == buffer.get(j).unwrap() {
                            index += 1;
                            continue 'chars;
                        }
                    }
                }
                break;
            }
        }
    }

    println!("Part 2: {}", index);
}
